std/typecons.d

```diff
@@ -7143,3 +7143,265 @@ unittest
+ *
+ * Params:
+ *     obj = the struct or class instance to access
+ *     name = the name of the field or property member
+ *     value = the new value to assign to the member
+ */
+void setField(Field, Obj, Str)(Obj obj, const Str name, auto ref Field value)
+    if ( is(Obj == class) && isSomeString!Str)
+{
+    enum refField = !is(Field == class) && (Field.sizeof > 16);
+    rtFieldDispatch!(true, Field, refField, Obj, false)(obj, name, value);
+}
+
+/// ditto
+void setField(Field, Obj, Str)(ref Obj obj, const Str name, auto ref Field value)
+    if (!is(Obj == class) && isSomeString!Str)
+{
+    enum refField = !is(Field == class) && (Field.sizeof > 16);
+    rtFieldDispatch!(true, Field, refField, Obj, true)(obj, name, value);
+}
+
+///
+@safe @nogc nothrow pure unittest
+{
+    struct Foo
+    {
+        int bar = 42;
+    }
+
+    Foo foo;
+    foo.setField("bar", 24);
+    assert(foo.bar == 24);
+}
+
+// mixin template for the get/setField functions
+private template rtFieldDispatch(bool set, Field, bool refField, Obj, bool refObj)
+{
+private:
+    enum propMix = (set ? `` : `return `) ~ `mixin("obj." ~ mName)` ~ (set ? ` = value; return;` : `;`);
+    enum rtMix = set ? `void` : (refField ? `ref Field` : `Field`);
+    enum objMix = (refObj ? (!set && refField ? `return ` : ``) ~ `ref ` : ``) ~ `Obj obj`;
+    enum fldMix = set ? `, ` ~ (refField ? `ref ` : ``) ~ `Field value` : ``;
+
+    enum dispMix = rtMix ~ ` rtFieldDispatch(Str)(` ~ objMix ~ `, const Str name` ~ fldMix ~ `)
+    {
+        ` ~ rtMix ~ ` checkType(const Str mName)() pure @safe
+        {
+            if (!__ctfe) assert (0);
+            ` ~ propMix ~ `
+        }
+
+        foreach (immutable Str mName; __traits (allMembers, Obj))
+        {
+            static if (__traits(compiles, checkType!mName()))
+            {
+                if (mName == name)
+                {
+                    ` ~ propMix ~ `
+                }
+            }
+        }
+
+        assert (0, "` ~ Obj.stringof ~ ` has no ` ~ (set ? `assignable` : (refField ? `lvalue ` : ``)) ~ `
+            property or field with the specified name matching ` ~ Field.stringof ~ `");
+    } `;
+
+public:
+    mixin(dispMix);
+}
+
+@safe nothrow pure unittest
+{
+    enum A =
+`   @safe: @nogc: nothrow: pure:
+    private:
+        int _valProp = 31;
+        int _refProp = -78;
+
+    public:
+        int field = 564;
+
+    @property:
+        int valProp() const
+        {
+            return _valProp;
+        }
+        void valProp(int val)
+        {
+            _valProp = val;
+        }
+
+        ref inout(int) refProp() inout
+        {
+            return _refProp;
+        }`;
+
+    static struct S
+    {
+        mixin(A);
+    }
+
+    static class C
+    {
+        mixin(A);
+    }
+
+    static void testRead(Obj)(const Obj obj) @nogc
+    {
+        assert(obj.getField!int("valProp") == 31);
+        assert(obj.getField!int("refProp"w) == -78);
```

Comment:
Shouldn't access to private attributes be only possible within the object itself?